mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,160 @@
+r""":mod:`mirgecom.artificial viscosity` Artificial viscocity for Euler.
+
+Artificial viscosity for the Euler Equations:
+
+.. math::
+
+    \partial_t \mathbf{Q} = \nabla\cdot{\varepsilon\nabla\mathbf{Q}}
+
+where:
+
+-  state $\mathbf{Q} = [\rho, \rho{E}, \rho\vec{V}, \rho{Y}_\alpha]$
+-  artifical viscosity coefficient $\varepsilon$
+
+To evalutate the second order derivative the problem is recast as a set of first
+ order problems:
+
+.. math::
+
+    \partial_t \mathbf{Q} = \nabla\cdot{\mathbf{R}}
+    \mathbf{R} = \varepsilon\nabla\mathbf{Q}
+
+where $\mathbf{R}$ is an intermediate variable.
+
+RHS Evaluation
+^^^^^^^^^^^^^^
+
+.. autofunction:: artificial_viscosity
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from pytools.obj_array import (
+    obj_array_vectorize,
+    obj_array_vectorize_n_args,
+)
+from meshmode.dof_array import thaw
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+from grudge.symbolic.primitives import TracePair
+from mirgecom.tag_cells import smoothness_indicator
+
+
+def _facial_flux_r(discr, q_tpair):
+
+    actx = q_tpair.int.array_context
+
+    flux_dis = q_tpair.avg
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    flux_out = flux_dis * normal
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
+
+
+def _facial_flux_q(discr, q_tpair):
+
+    actx = q_tpair[0].int.array_context
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    flux_out = np.dot(q_tpair.avg, normal)
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
+
+
+def artificial_viscosity(discr, t, eos, boundaries, r, alpha, **kwargs):
+    r"""Compute artifical viscosity for the euler equations."""
```

Comment:
This interface should be documented.  [Here's a similar one](https://github.com/illinois-ceesd/mirgecom/blob/171e59fe483f11448b4d9967dd10a3512753bbcf/mirgecom/euler.py#L279) to give you some ideas about what to write for some of your arguments.

Reply:
Added interface parameter documentation: https://github.com/illinois-ceesd/mirgecom/pull/213/commits/218cb523e501a27c1aabb6ec56899efb9975a816